Add draft of Array._repr_html_

mrocklin opened the pull request:
This provides more information about Dask arrays as an HTML repr

Currently it includes a table of information (bytes, shape, dtype, ...)
A visual representation of the grid in 1d or 2d (working on 3d next)

A rendered result is availble here https://nbviewer.jupyter.org/urls/gist.githubusercontent.com/mrocklin/0df34e9d609f38a2bf3df9311909bb2c/raw/7ab5513cc1c6740ca8b2f4785ecd7d3a3869012c/dask-array-repr.ipynb

<img width="816" alt="Screen Shot 2019-05-11 at 11 08 53 AM" src="https://user-images.githubusercontent.com/306380/57572245-4428c500-73dd-11e9-9d72-ed889ad82a80.png">


### TODO
-  3d
-  handle greater dimensions by having a few grids displayed .  So for example 4d would be a 3d grid followed by a 1d grid
-  Improve styling and placement of grid (visual design help most welcome)
-  Figure out better heuristics on how to handle many chunks or high degrees of asymmetry

cc @shoyer @birdsarah @rabernat @jhamman 

My hope is that with changes like these (I plan to do something similar for high level graphs) we can improve users' literacy and ability to reason about their computations.

shoyer commented:
On my screen, the text in the SVG showing dimension sizes looks really bold, with overlapping letters:
![image](https://user-images.githubusercontent.com/1217238/57572465-7501fd80-73cf-11e9-949e-c44f579e76e4.png)

shoyer commented:
Maybe instead of "(3, 10) → 30" for # Chunks, this could be "3 x 10 = 10"?

mrocklin commented:
Yes, on mine as well.  Does anyone know enough about SVG to help here?  I don't know very well how to control font styling.

shoyer commented:
I think you might be able to do this in pure CSS, which would likely render more consistently. That's just a guess, though, I'm definitely not an expert here.

shoyer commented:
To be honest, I'm not sure I love the big table and 2D representation of arrays. It feels much lower information density than the old repr. My preference would be to try to squeeze the HTML repr into roughly 4-5 lines of text, which could perhaps be achieved by splitting the table into more columns and using a smaller visual depiction.

I guess there's a tradeoff between representations optimized for new users and for people who understand the dask.array data model. If I were optimizing this for experienced users, I would consider something like the following, with a barplot showing dimension/chunk size for each axis on a log-scale:
![image](https://user-images.githubusercontent.com/1217238/57572679-987a7780-73d2-11e9-9eef-fbc9ba44344e.png)

shoyer commented:
I do really like the idea of exploring HTML reprs, though!

birdsarah commented:
On the advanced vs beginner thing, I agree. Or even the difference between when you're working and learning and when it's finally working and you want to have a compact and complete notebook. I'm thinking maybe something like sympy's init_printing function that we can recommend in docs and then it's easy to opt out of - or even to set by environment.

birdsarah commented:
I also think there's value in building custom visualizations for 2, 3, and 4D data which are all very common use cases - especially for beginners. And then have a more generic representation for the higher dimensional data.

mrocklin commented:
I find the grid structure to be helpful when playing around with dask arrays.  Even as an expert I found myself relying on it almost immediately when I started playing around in the notebook.  Previously I kept a mental model of chunks in my head, and now I found myself being lazy and leaving that to the UI.  

For new users I would like for the repr to be interpretable without explanation, which, to me, rules out the histograms.  

However, adding the constraint of information density should help I think.  I'm hoping that if we're clever that we can come up with something that is both compact and easily interpretable without explanation.  We don't need to visually show every chunk in the array.  I think that there are diminishing returns after showing a few representative chunks per dimension.  I think that some sort of gridded structure is intutitive, but I think that it breaks down with a large number of chunks.

mrocklin commented:
> I do really like the idea of exploring HTML reprs, though!

Yes!  I was inspired by Iris actually (cc @jacobtomlinson).  More PyData projects should explore HTML reprs.  Turns out that our displays are capable of more than just fixed-width formatted text :)

mrocklin commented:
> I also think there's value in building custom visualizations for 2, 3, and 4D data which are all very common use cases - especially for beginners. And then have a more generic representation for the higher dimensional data.

I'm generally on board with this (assuming we go for something gridded like what is proposed here rather than a histogram approach, where this problem goes away).

Thoughts on what 3D or 4D should look like?

shoyer commented:
I wonder if log or power-law scaling for sizes would work even with 2D arrays?

mrocklin commented:
A small test to encode some expectations explicitly (code not yet written)

```python
def grid_points(chunks):
    """ Compute locations of lines for drawing a grid """

def test_draw_grid_points():
    x = da.ones((100, 100), chunks=(10, 10))
    x, y = grid_points(x.chunks)
    assert (x == y).all()  # respects symmetry

    x = da.ones((10, 50), chunks=(10, 10))
    x, y = grid_points(x.chunks)
    assert len(x) == 2
    assert len(y) == 5   # happy to draw modest grids explicitly
    assert y[-1] > x[-1]  # larger axes should be larger

    x = da.ones((10, 10000), chunks=(10, 100))
    x, y = grid_points(x.chunks)
    assert len(x) == 2
    assert 5 <= len(y) < 30  # not too many lines
    assert y[-1] > x[-1]  # still a difference
    assert y[-1] < x[-1] * 20  # but not beyond perception
```

mrocklin commented:
Log or power-law scaling makes sense.  We might also go more extreme and just cap the maximum aspect ratio at something like 10x.  Differences between the axes might not matter much beyond that.

dhirschfeld commented:
Maybe a `repr_html(**kwargs)` function with kwargs which can control the format of the output with `_repr_html`_ defaulting to a compact version?

e.g. `repr_html(show_grid=False, **other_useful_kwargs)`

shoyer commented:
One nice thing about log scaled barplots is they correspond nicely with how
dimension sizes combine: if two different barplots each add up to the same
total area, then the arrays have the same size, because
log(x)+log(y)=log(x*y). So I imagine you could pretty quickly build up an
intuition for what "good" chunk sizes look like at a glance.

On Sat, May 11, 2019 at 8:51 PM Dave Hirschfeld <notifications@github.com>
wrote:

> Maybe a repr_html(**kwargs) function with kwargs which can control the
> format of the output with _repr_html_ defaulting to a compact version?
>
> e.g. repr_html(show_grid=False, **other_useful_kwargs)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/4794#issuecomment-491562234>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAJJFVWAZZM6XEYKBK26FLLPU6H4RANCNFSM4HMIOSPA>
> .
>

mrocklin commented:
Now with 3d and logarithmic scaling

https://nbviewer.jupyter.org/urls/gist.githubusercontent.com/mrocklin/0df34e9d609f38a2bf3df9311909bb2c/raw/15d41f26800710acc63daabf7e80bfda703b883d/dask-array-repr-2.ipynb

<img width="530" alt="Screen Shot 2019-05-12 at 8 38 46 PM" src="https://user-images.githubusercontent.com/306380/57591227-fbb0fa80-74f5-11e9-89c2-feb66a92d069.png">

mrocklin commented:
We could also make things more compact.

## 200px size
<img width="530" alt="Screen Shot 2019-05-12 at 8 38 46 PM" src="https://user-images.githubusercontent.com/306380/57591303-54809300-74f6-11e9-9e1f-de836a845036.png">

## 120px size
<img width="454" alt="Screen Shot 2019-05-12 at 8 40 54 PM" src="https://user-images.githubusercontent.com/306380/57591288-46cb0d80-74f6-11e9-9a41-71ad9854e546.png">

martindurant commented:
I find that the array name is not very useful, since you can never read it. It could be removed, or be more of a "title", spanning across the table and graphic. 

Does this extend to
- zero dimensions
- >3 dimensions
- complex dtypes?

mrocklin commented:
See conversation above and example notebook for dimensions other then 1,2,3 (the answer is no, it doesn't)

We just print the dtype, so I don't see any reason to suspect that it wouldn't work.

Another special case you haven't mentioned is unknown chunks.

mrocklin commented:
Now with lighter weight dimension text

<img width="419" alt="Screen Shot 2019-05-12 at 8 58 31 PM" src="https://user-images.githubusercontent.com/306380/57591783-b6da9300-74f8-11e9-9152-a1fe86dcf51c.png">

jacobtomlinson commented:
All credit for the [Iris repr](https://github.com/SciTools/iris/pull/2918#issuecomment-387413538) goes to @DPeterK. These HTML reprs are awesome though, make life much easier and we should repr all the things! I did the [worker logs in dask-kubernetes](https://github.com/dask/dask-kubernetes/pull/142) and it makes it so much easier to debug issues.

It would be interesting to think about nesting these reprs too, or at least making reusable components. I commonly use Iris and Xarray which both use dask array as the internal array. It would be nice for the HTML repr for both of those to pull out at least some of the dask array one!

mrocklin commented:
I've added some modest testing and am now raising informative errors.  I've also added in docstrings in some of the functions more likely to be seen by users or that are somewhat complex.

shoyer commented:
Xarray has an HTML repr in progress, but it needs a bit more work to bring
it across the finish line: https://github.com/pydata/xarray/pull/1820

On Mon, May 13, 2019 at 7:07 AM Matthew Rocklin <notifications@github.com>
wrote:

> I've added some modest testing and am now raising informative errors. I've
> also added in docstrings in some of the functions more likely to be seen by
> users or that are somewhat complex.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/4794?email_source=notifications&email_token=AAJJFVR2B3SZUNFWGMHKG63PVFYZBA5CNFSM4HMIOSPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVINO4Q#issuecomment-491837298>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJJFVSSKL2GCMSZNCKAFEDPVFYZBANCNFSM4HMIOSPA>
> .
>

mrocklin commented:
Ooh, that's nice

On Mon, May 13, 2019 at 9:36 AM Stephan Hoyer <notifications@github.com>
wrote:

> Xarray has an HTML repr in progress, but it needs a bit more work to bring
> it across the finish line: https://github.com/pydata/xarray/pull/1820
>
> On Mon, May 13, 2019 at 7:07 AM Matthew Rocklin <notifications@github.com>
> wrote:
>
> > I've added some modest testing and am now raising informative errors.
> I've
> > also added in docstrings in some of the functions more likely to be seen
> by
> > users or that are somewhat complex.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/dask/dask/pull/4794?email_source=notifications&email_token=AAJJFVR2B3SZUNFWGMHKG63PVFYZBA5CNFSM4HMIOSPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVINO4Q#issuecomment-491837298
> >,
> > or mute the thread
> > <
> https://github.com/notifications/unsubscribe-auth/AAJJFVSSKL2GCMSZNCKAFEDPVFYZBANCNFSM4HMIOSPA
> >
> > .
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/4794?email_source=notifications&email_token=AACKZTDE5VSIYZNDWYGGSBDPVF4H5A5CNFSM4HMIOSPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVIQM4Y#issuecomment-491849331>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACKZTAPT5RUUU3LVEDWI6LPVF4H5ANCNFSM4HMIOSPA>
> .
>

mrocklin commented:
> I find that the array name is not very useful, since you can never read it. It could be removed, or be more of a "title", spanning across the table and graphic.

I wanted to have a few characters from the hash, just so that people can identify differences, but you're right that this probably isn't critical or even really useful to most users.  Is the name (`ones` in this case)?

mrocklin commented:
What else should be done here?  @shoyer any further suggestions?  Is the size small/compact enough for you?  Are there ways that you can recommend making things more information dense?

shoyer commented:
@mrocklin can you update the notebook with the latest output from this PR?

mrocklin commented:
Sure.
https://nbviewer.jupyter.org/urls/gist.githubusercontent.com/mrocklin/0df34e9d609f38a2bf3df9311909bb2c/raw/ad757237590c4bd0e3141f937d7b9aeefec06c64/dask-array-repr-3.ipynb

On Mon, May 13, 2019 at 1:03 PM Stephan Hoyer <notifications@github.com>
wrote:

> @mrocklin <https://github.com/mrocklin> can you update the notebook with
> the latest output from this PR?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/4794?email_source=notifications&email_token=AACKZTFUKJBAIMHFAPJ2FCLPVGUOHA5CNFSM4HMIOSPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVJDCEI#issuecomment-491925777>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACKZTBZ7LHL3HX6NB5EIVTPVGUOHANCNFSM4HMIOSPA>
> .
>

mrocklin commented:
It would be good to get a sense of what happens when we apply this to real
datasets (cc @rabernat)

On Mon, May 13, 2019 at 1:05 PM Matthew Rocklin <mrocklin@gmail.com> wrote:

> Sure.
> https://nbviewer.jupyter.org/urls/gist.githubusercontent.com/mrocklin/0df34e9d609f38a2bf3df9311909bb2c/raw/ad757237590c4bd0e3141f937d7b9aeefec06c64/dask-array-repr-3.ipynb
>
> On Mon, May 13, 2019 at 1:03 PM Stephan Hoyer <notifications@github.com>
> wrote:
>
>> @mrocklin <https://github.com/mrocklin> can you update the notebook with
>> the latest output from this PR?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/dask/dask/pull/4794?email_source=notifications&email_token=AACKZTFUKJBAIMHFAPJ2FCLPVGUOHA5CNFSM4HMIOSPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVJDCEI#issuecomment-491925777>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AACKZTBZ7LHL3HX6NB5EIVTPVGUOHANCNFSM4HMIOSPA>
>> .
>>
>

shoyer commented:
This is looking pretty good to me.

I still imagine the "bar plot" representation (https://github.com/dask/dask/pull/4794#issuecomment-491527354) could be nice as a quick reference for expert users -- it usable even if it's pretty small. That said, this could definitely go in later.

For saving space, what about splitting the table into two lists of parameters? Could keep the stuff in the repr on the left (shape, dtype, chunks) and put the extra derived info in another column on the right. You could potentially play around with different formatting, too, to make it clear that the derived info is different from the fields that are directly part of dask's data model.

andersy005 commented:
> It would be good to get a sense of what happens when we apply this to real
datasets 

@mrocklin, I am not 100% sure this is what you are looking for, but, here's what I get for two different datasets (3D, and 4D)

https://nbviewer.jupyter.org/gist/andersy005/c0434e0cec5a7625ae2849e1d954b542

![Screen Shot 2019-05-13 at 1 04 55 PM](https://user-images.githubusercontent.com/13301940/57647299-d23cb100-757f-11e9-8115-78792e114035.png)
![Screen Shot 2019-05-13 at 1 04 46 PM](https://user-images.githubusercontent.com/13301940/57647300-d2d54780-757f-11e9-9231-bd2e723761d4.png)

mrocklin commented:
Thanks @andersy005 !  That's useful to see

I guess one question would be "how much do we care about handling greater than 3 dimensions".  For Xarray/Iris users the answer here might be "a lot".  For this we can do a few 1-2-3D arrays arranged horizontally (I have a bit of code to do this now) or something more generic like @shoyer's histogram plots.

mrocklin commented:
Now with ndarray support-ish

<img width="680" alt="Screen Shot 2019-05-13 at 3 02 20 PM" src="https://user-images.githubusercontent.com/306380/57650309-251e6480-7590-11e9-9548-0425b9b57310.png">

rabernat commented:
This looks amazing! Thanks for making it. Will be so useful for teaching.

mrocklin commented:
> For saving space, what about splitting the table into two lists of parameters? Could keep the stuff in the repr on the left (shape, dtype, chunks) and put the extra derived info in another column on the right. You could potentially play around with different formatting, too, to make it clear that the derived info is different from the fields that are directly part of dask's data model.

So, question for the crowd, do we want the repr to be smaller?  Currently the table and the grid image are about the same size (120px).  We can shrink them, or grow them, or keep them the same size,  I'm currently inclined towards keeping things as they are currently.   I don't think I would want to shrink the grid image much more than it is now.

dhirschfeld commented:
The size in bytes of each-chunk might be a useful metric? 
i.e. the user may want to ensure each chunk fits in the cache

jakirkham commented:
In case it is of interest, here is a similar feature in Zarr.

ref: https://nbviewer.jupyter.org/github/zarr-developers/zarr/blob/master/notebooks/repr_info.ipynb

cc @alimanfoo

mrocklin commented:
OK, handling @martindurant 's feedback here: 

>  I find that the array name is not very useful, since you can never read it. It could be removed, or be more of a "title", spanning across the table and graphic.

And @dhirschfeld 's feedback here:

> The size in bytes of each-chunk might be a useful metric?
> i.e. the user may want to ensure each chunk fits in the cache

We now have the following:

<img width="539" alt="Screen Shot 2019-05-13 at 7 31 19 PM" src="https://user-images.githubusercontent.com/306380/57662748-7a6d6c80-75b6-11e9-906c-d2f40bdeb919.png">

mrocklin commented:
I'm inclined to merge soonish if there are no objections.  This is the sort of thing that we can modify easily in the future.  There aren't any strong contracts to break here.

martindurant commented:
I like this last version better. I am uncertain whether there's anything to be gained from showing the array name, except that it is the thing that shows up most prominently in progress-bars and on the dashboard.

stsievert commented:
Ditto, I’m also a fan on the two column version. It provides a clean separation and shows more information than https://github.com/dask/dask/pull/4794#issuecomment-491965341 in fewer rows (4 vs 6 rows).

alimanfoo commented:
Really awesome to see this! It will help hugely for gaining understanding and intuitions for how dask works. (Numpy really should do something similar :-)

Don't know if anyone else is feeling this, but one thing I'm finding a little confusing at the moment is that how the dimensions are displayed changes depending on the number of dimensions. E.g., if you have a 2D array then the first dimension is displayed as vertical length. But then if you have a 3D array then the first dimension is displayed as depth. 

One thing I find in explaining arrays to others is that it helps to think of the first dimension visually in the same way and always "draw" it the same way. I.e., when I draw an array, I always draw the first dimension as vertical, the second dimension as horizontal, and then if there is a third dimension, that becomes depth. I don't really mind which way round, as long as it's consistent.

alimanfoo commented:
Btw wondering if we could factor this out somehow. Dask, NumPy, Zarr, ... could all use the same HTML repr very effectively. Only difference would be the rows in the table on the left.

DPeterK commented:
@mrocklin this is great! Here's a screenshot of trying this on the dask array at the core of a (small) 4D Iris cube (bigger, more interesting cube incoming):

![Screenshot 2019-05-14 at 11 42 33](https://user-images.githubusercontent.com/3473068/57692033-a4b63d00-763d-11e9-897e-5a2ee049b361.png)

The one odd thing is the pretty repr seems to have generated a _5th_ dimension (and called it `L`), which doesn't exist in the cube's dask array.

jacobtomlinson commented:
Something kind of like this would be nice

![image](https://user-images.githubusercontent.com/1610850/57697699-bacafa00-764b-11e9-8b24-e62b2379b2a3.png)

DPeterK commented:
Here's a promised, _much_ larger dataset loaded as a cube:

![Screenshot 2019-05-14 at 13 51 57](https://user-images.githubusercontent.com/3473068/57699382-83f6e300-764f-11e9-80ec-8954d6f8f79c.png)

mrocklin commented:
From @alimanfoo 

> Don't know if anyone else is feeling this, but one thing I'm finding a little confusing at the moment is that how the dimensions are displayed changes depending on the number of dimensions

Yes, I originally kept things consistent so that the third dimension was the depth dimension.  I found that it conflicted with my intuition of what arrays looked like, possibly due to a history of imaging datasets.  I wanted the fastest-moving dimensions (like (100, **1024**, **1024**)) to be facing me as a a rectangle, and then I wanted the slowest moving dimension going away as depth.  I fully acknowledge though that this may not be ideal.  It would be good to get others' opinions on how best to arrange the dimensions.

mrocklin commented:
@DPeterK @jacobtomlinson note that you can get just the grid without the table by using the `to_svg` method.  You can also control the size of the image there with the `size=` keyword.

mrocklin commented:
I think that the only pending issue is @alimanfoo 's from above about which dimension goes where.  If people have thoughts that they're able to share that would be great.  Additional perspectives would be valuable here.

shoyer commented:
I would lean towards including the array name in the HTML repr, under the model that the HTML repr should contain a strict super-set of the information in the text repr. Unless there's something I'm missing about different use-cases for these?

alimanfoo commented:
Heh, I can see this is hard to generalise as different domains think about
arrays in different ways. For our genomic data, array shape is typically
(100 million, 10000, 2) so putting the first dimension as depth will look
pretty weird :-)

Maybe you could do a webgl version which the user can rotate in 3d to their
liking :-)

On Tue, 14 May 2019, 13:57 Matthew Rocklin, <notifications@github.com>
wrote:

> From @alimanfoo <https://github.com/alimanfoo>
>
> Don't know if anyone else is feeling this, but one thing I'm finding a
> little confusing at the moment is that how the dimensions are displayed
> changes depending on the number of dimensions
>
> Yes, I originally kept things consistent so that the third dimension was
> the depth dimension. I found that it conflicted with my intuition of what
> arrays looked like, possibly due to a history of imaging datasets. I wanted
> the fastest-moving dimensions (like (100, *1024*, *1024*)) to be facing
> me as a a rectangle, and then I wanted the slowest moving dimension going
> away as depth. I fully acknowledge though that this may not be ideal. It
> would be good to get others' opinions on how best to arrange the dimensions.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/4794?email_source=notifications&email_token=AAFLYQTJE627ZXANGWRHHULPVKZNJA5CNFSM4HMIOSPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVLMNPI#issuecomment-492226237>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFLYQQR4WPQXSC5VZZVVD3PVKZNJANCNFSM4HMIOSPA>
> .
>

mrocklin commented:
> Maybe you could do a webgl version which the user can rotate in 3d to their liking :-)

Sounds like great future work :)

mrocklin commented:
> I would lean towards including the array name in the HTML repr, under the model that the HTML repr should contain a strict super-set of the information in the text repr. Unless there's something I'm missing about different use-cases for these?

@martindurant gave the opposite recommendation in https://github.com/dask/dask/pull/4794#issuecomment-491649895 on the grounds that it's not particularly useful.  

> under the model that the HTML repr should contain a strict super-set of the information in the text repr

I don't see a strong motivation for this personally.  I'm optimizing the HTML repr mostly for user experience rather than comprehensive representation.

martindurant commented:
> gave the opposite recommendation

I suggested that just seeing part of the name wasn't interesting; if you wanted to have a superset of the text repr, then you would want the whole name, which is too long for the previous location. I could imagine it as a heading for the HTML block.

stsievert commented:
Would showing the text repr in the title-text would help both https://github.com/dask/dask/pull/4794#issuecomment-492715663 and https://github.com/dask/dask/pull/4794#issuecomment-492357103? Something like

``` html
<table>
  <tr title="ones-353699120031">
    <th></th>
    <th>Array</th>
    <th>Chunks</th>
  </tr>
  ...
</table>
```

This looks something like

<img width="410" alt="Screen Shot 2019-05-15 at 11 25 58 AM" src="https://user-images.githubusercontent.com/1320475/57792301-34d3af80-772e-11e9-9122-998f244909b3.png">

shoyer commented:
I would avoid title text. You can't select it for copy/paste, and it's not something notebook users know to look for.

mrocklin commented:
The array name seems contentious enough that I think I'd like to leave it for future work.  I don't think it's necessary to hold up the rest of this PR while waiting for folks to come to an agreement there.

If there are no objections I'll plan to merge this tomorrow.

mrocklin commented:
Thanks for the review all.  I'm quite happy with how this turned out.  It was much nicer for all of the added perspectives.
